Add job script validations

AriettaZ opened the pull request:
### Overview
- [x] Add job script validation and provide recommendation for users
- [x] Step1: Omit files larger than 65KB for this field
- [x] Step2: Group files to "suggested" and "others" and sort alphabetically


 - Suggested files meet any of the following criteria
                        1. File name extension is one of `[".sh", ".job", ".slurm", ".batch", ".qsub", ".sbatch", ".srun", ".bsub"]`
                            2.  First line of the file is a shebang line
                              3.  First 1000 bytes of the file contain '#PBS' or '#SBATCH"
- Others files are files not omitted in Step1 and not meet requirement to be suggested

### Tests
- [x] Tested the example on the issue page
- [x] Tested the job script under the sub folder
![dropdown-with-2-groups](https://user-images.githubusercontent.com/22674713/68895237-e4e29f00-06f6-11ea-9f48-1da254ce8b80.JPG)



Fixes #105

ericfranz commented on apps/myjobs/app/models/workflow_helper.rb:
```diff
@@ -0,0 +1,52 @@
+class WorkflowHelper
```
I know I suggested this name, but I now realize that `WorkflowHelper` could be confused with `WorkflowHelpers`.

One refactor suggestion would be to turn this into a `WorkflowFile` object instead, whose single attribute is a path to the file.

Then the class might look something like this:

```ruby
class WorkflowFile
  attr_reader :path

  def initialize(path)
    @path = path
  end

  def valid_script?
    @valid_script ||= valid_size?
  end

  def suggested_script?
  end

  def valid_size?
    File.size(path).to_f/1024 <= 65
  end
```

Then on `Workflow` you could add:

```ruby
def files
  if File.directory?(staged_dir)
    Find.find(staged_dir).drop(1).select {|f| File.file?(f) }.map {|f| WorkflowFile.new(f) }
  else
    []
  end
end
```

ericfranz replied:
Or don't do `.select {|f| File.file?(f) }` and then do that in WorkflowFile (with the logic you have now - if it isn't a File it can't be a valid script.

AriettaZ replied:
The new commit fixes this feedback by adding a new WorkflowFile model and removing the WorkflowHelper

ericfranz commented on apps/myjobs/app/models/workflow.rb:
```diff
@@ -124,14 +124,46 @@ def stage
+  # Then, map files to WorkflowFile objects
+  #
+  # @return [WorkflowFile] An array of WorkflowFile Objects of all the files of a directory
   def folder_contents
```
could cache this like:

```ruby
def folder_contents
  return @folder_contents if defined? @folder_contents

  ...
end
```

ericfranz replied:
This actually only works if you set the @folder_contents otherwise. For example:

```ruby
def folder_contents
  return @folder_contents if defined? @folder_contents

  if File.directory?(self.staged_dir)
    @folder_contents = Find.find(self.staged_dir).drop(1).select {|f| File.file?(f) }.map {|f| WorkflowFile.new(f, self.staged_dir)}
  else
    @folder_contents = []
  end
end
```

ericfranz commented on apps/myjobs/app/models/workflow.rb:
```diff
@@ -124,14 +124,46 @@ def stage
+      end
+    }
+    [["Suggested files", suggested_files], ["Others", other_valid_files]]
+  end
```
Is this more readable? I don't know:

```ruby
def grouped_script_options
  {
    "Suggested files" => folder_contents.select(&:suggested_script?).map { |f| [f.relative_path, f.path] },
    "Others" => folder_contents.reject(&:suggested_script?).map { |f| [f.relative_path, f.path] },
  }.to_a
end
```

You can also keep as is which is more efficient but use `.each` here not `.map`.

AriettaZ replied:
Thanks! It looks better than what I have! And I just realized `grouped_options_for_select` can take in either a hash or an array. So my new commit doesn't have `to_a`.

ericfranz commented on apps/myjobs/app/models/workflow_file.rb:
```diff
@@ -0,0 +1,47 @@
+  
+  # Return true if file starts with a shebang line
+  def starts_with_shebang_line?
+    (File.open(path) { |f| f.read(2) }) == "#!"
```
Test if a file has permissions 000 or something funny like that and make sure this doesn't break.

AriettaZ replied:
Good call! It does raise file permission errors. I added rescue to the function.

ericfranz commented on apps/myjobs/app/views/workflows/show.json.jbuilder:
```diff
@@ -3,16 +3,17 @@ json.set! 'status_label', status_label(@workflow)
-  json.set! 'type', Pathname(content).file? ? 'file' : 'dir'
-  json.set! 'fsurl', Filesystem.new.fs(content)
-  json.set! 'fs_base', Filesystem.new.fs(File.dirname(content))
+json.folder_contents (@workflow.folder_contents).each do |content|
```
Since `folder_contents` return `WorkflowFile` objects, maybe the variable `content` should be replaced by `file`, then `file.path` reads better than `content.path`.

AriettaZ commented:
Thank you for your feedback! I cleaned up the code and updated the variables' names.

One issue existed is that the titles for option groups will still be displayed when no options available under that group. It's handled by removing empty option groups. 

1. Have both suggested and other files:
![both valid and suggested](https://user-images.githubusercontent.com/22674713/69452459-c05a7880-0d2f-11ea-878c-b791a77c7f3b.JPG)
2. Only have suggested files:
![only suggested](https://user-images.githubusercontent.com/22674713/69452831-a79e9280-0d30-11ea-838d-b1da5c32949d.JPG)
3. Have valid but not suggested files:
I changed the group title "Others" to "Other valid file(s)" because "Others" could be confusing when it's the only group available. We can also remove the "Other valid file(s)"  title completely if that's more user-friendly. 
![only other files](https://user-images.githubusercontent.com/22674713/69452917-dfa5d580-0d30-11ea-8df3-452217268285.JPG)

4. No valid files or folder is empty: 
![empty folder](https://user-images.githubusercontent.com/22674713/69452775-82118900-0d30-11ea-9f4c-db78cdaad9d0.JPG)

ericfranz commented on apps/myjobs/app/models/workflow.rb:
```diff
@@ -124,14 +124,43 @@ def stage
+    group_options = {
+      "Suggested file(s)" => folder_contents.select(&:suggested_script?).map { |f| [f.relative_path, f.path] },
+      "Other valid file(s)" => folder_contents.select(&:valid_script?).reject(&:suggested_script?).map { |f| [f.relative_path, f.path] },
+    }.reject {|k,v| v.blank?}
```
Since value is an array, `v.blank?` would read better as `v.empty?`.

ericfranz commented on apps/myjobs/app/models/workflow_file.rb:
```diff
@@ -0,0 +1,55 @@
+  end
+
+  # Return true if first 1000 bytes of file contain '#PBS' or '#SBATCH"
+  def has_specific_command?
```
Better to rename `has_specific_command?` to `has_resource_manager_directive?`. Also based on https://slurm.schedmd.com/rosetta.pdf these are all the script directives we should search for:

- `#PBS`
- `#SBATCH`
- `#BSUB`
- `#$`

And maybe open an issue in ood_core to add to job adapters a method to return the "script_directive_prefix" which will return these strings. Because this is an example of something that should ultimately be handled by ood_core.

AriettaZ commented:
Thank you for the feedback and help identify the script directives! I updated `v.blank?` to `v.empty?`, added the checks for the new resource manager directives, and opened the issue you mentioned: https://github.com/OSC/ood_core/issues/161.

MorganRodgers commented on apps/myjobs/app/models/workflow_file.rb:
```diff
@@ -0,0 +1,55 @@
+  def has_resource_manager_directive?
+    begin
+      contents = File.open(path) { |f| f.read(1000) }
+      contents && (contents.include?("#PBS") || contents.include?("#SBATCH") || contents.include?('#BSUB') || contents.include?('#$'))
```
For the `contents.include?` predicate I would prefer:

```ruby
['#PBS', '#SBATCH', '#BSUB', '#$'].any?{ |directive| contents.include?(directive) }
```

MorganRodgers replied:
Although this is fine as is: a better design here would be for `ood_core`'s Adapters to provide their directive as a constant and for this array to be built from the available clusters. The more tests we have the more likely we are to get a false positive.

MorganRodgers replied:
Just noticed https://github.com/OSC/ood_core/issues/161. +1

MorganRodgers commented on apps/myjobs/app/models/workflow_file.rb:
```diff
@@ -0,0 +1,55 @@
+
+  # Return true if file size is smaller than 65KB
+  def valid_size?
+    File.size(path).to_f/1024 <= 65
```
I argue that 65KB is an arbitrary size, and as such we should not hard code it. This should be configurable.

MorganRodgers commented on apps/myjobs/app/models/workflow.rb:
```diff
@@ -255,7 +285,7 @@ def stop_machete_jobs(jobs)
     end
   end
-
+  
```
Pointless whitespace...
